<a href="https://colab.research.google.com/github/crisbpadilla/CapstoneProject-1-M2M/blob/main/CapstoneProject_2_Regression.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Machine learning- compressive strenght prediction modeling
  In the present capston project, the same data used in the first project will be scaled and transform to preprocess the data, prepared the training and testing data and evaluate the different regression models that could be used to predict the compressive strength of the concrete mix.

  So, the dataset was obtain from kaggle:

In [31]:
import pandas as pd
import numpy as np

concrete=pd.read_csv('concrete.csv')

In the first capston project it was possible to identify that the main components that helps to improve the development and higher values of compressive strength were the SCM and the superplastice.
So due this, we will firstly checked the correlation matrix between the variables to have  a better idea of it and see if this could help to take desicions in the future models.

In [32]:
from bokeh.plotting import figure, output_notebook, show, output_notebook
from bokeh.models import ColumnDataSource, LinearColorMapper,ColorBar

output_notebook()
material_corr=concrete.corr()

material_corr = material_corr.stack().reset_index()
material_corr.columns = ['x', 'y', 'value']

material_corr['value'] = material_corr['value'].round(2)

source = ColumnDataSource(material_corr)

# Define color mapper
color_mapper = LinearColorMapper(palette="Inferno256", low=-1, high=1)

# Create figure
p = figure(title="Correlation Heatmap", x_range=list(concrete.columns), y_range=list(concrete.columns),
           x_axis_location="above", width=600, height=600,
           tools="hover", tooltips="@x, @y: @value{0.2f}")

p.rect(x="x", y="y", width=1, height=1, source=source,
       fill_color={'field': 'value', 'transform': color_mapper}, line_color=None)

# Add text annotations
from bokeh.models import LabelSet
labels = LabelSet(x='x', y='y', text='value', source=source,text_align='center',
                  text_baseline='middle',text_font_size='8pt',text_color='white')
p.add_layout(labels)

# Add color bar
color_bar = ColorBar(color_mapper=color_mapper, label_standoff=12, border_line_color=None, location=(0, 0))
p.add_layout(color_bar, 'right')

p.xaxis.major_label_orientation = 0.75

# Show plot
show(p)


Reviewing the heatmap it is possible to observe the inverse correlation between water an superplastic, variable that could be flattened as they are both lightly correlated ( superplastic is used to speed up the compressive strength gain and the reduction of water in the mix).
 So we will check the performance of the model, reducing the demionsionality of this model , taking water and superplastic into one unique component.

# Preprocessing data

In [33]:
print(concrete.head())
print(concrete.info())

   cement   slag    ash  water  superplastic  coarseagg  fineagg  age  \
0   141.3  212.0    0.0  203.5           0.0      971.8    748.5   28   
1   168.9   42.2  124.3  158.3          10.8     1080.8    796.2   14   
2   250.0    0.0   95.7  187.4           5.5      956.9    861.2   28   
3   266.0  114.0    0.0  228.0           0.0      932.0    670.0   28   
4   154.8  183.4    0.0  193.3           9.1     1047.4    696.7   28   

   strength  
0     29.89  
1     23.51  
2     29.22  
3     45.85  
4     18.29  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1030 entries, 0 to 1029
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   cement        1030 non-null   float64
 1   slag          1030 non-null   float64
 2   ash           1030 non-null   float64
 3   water         1030 non-null   float64
 4   superplastic  1030 non-null   float64
 5   coarseagg     1030 non-null   float64
 6   fineagg       1030 non

the dataframe is build by the raw materials that composed the concrete mix  (kg/m^3), age (days) and compressive strength of the sample.

with the idea of see the impact of adding   categorical values to the model, depending the type of mix, the next code includes the next categories.

    * all mix
    * slag
    * ash
    * superplastic
    * slag-ash
    * slag-superplastic
    * ash-superplastic.
    * just cement


In [34]:
concrete_pluscat=concrete.copy()
# cuales son los escenarios? 9 (matriz 3x3) , la mitad menos 1 ~ 3
#a) slag-ash b)slag-superplastic c)ash-superplastic

concrete_pluscat['type_mix'] = 'all mix'
concrete_pluscat.loc[
    (concrete_pluscat['slag']>0) & (concrete_pluscat[['superplastic','ash']] == 0).all(axis=1),
    'type_mix',
] = 'slag'
concrete_pluscat.loc[
    (concrete_pluscat['ash']>0) & (concrete_pluscat[['superplastic','slag']] == 0).all(axis=1),
    'type_mix',
] = 'ash'
concrete_pluscat.loc[
    (concrete_pluscat['superplastic']>0) & (concrete_pluscat[['slag','ash']] == 0).all(axis=1),
    'type_mix',
] = 'superplastic'

concrete_pluscat.loc[
    (concrete_pluscat[['slag', 'ash']].gt(0).all(axis=1)) & (concrete_pluscat['superplastic'] == 0),
    'type_mix',
] = 'slag-ash'

concrete_pluscat.loc[
    (concrete_pluscat[['slag', 'superplastic']].gt(0).all(axis=1)) & (concrete_pluscat['ash'] == 0),
    'type_mix',
] = 'slag-superplastic'

concrete_pluscat.loc[
    (concrete_pluscat[['ash', 'superplastic']].gt(0).all(axis=1)) & (concrete_pluscat['slag'] == 0),
    'type_mix',
] = 'ash-superplastic'
concrete_pluscat.loc[
    (concrete_pluscat['cement']>0) & (concrete_pluscat[['superplastic','slag','ash']] == 0).all(axis=1),
    'type_mix',
] = 'just cement'

In [35]:
concrete_pluscat.groupby('type_mix')['strength'].count()

type_mix
all mix              225
ash                    6
ash-superplastic     233
just cement          209
slag                 164
slag-superplastic    170
superplastic          23
Name: strength, dtype: int64

there is no slag-ash mix. giving 6 dummie columns dropping the first one.

###Preparing data for modelling

  As was seen in the first capstone project, the there is data distributed in all the categories related with the type of mix before the 100 days (seen in the next graph)

In [36]:
from bokeh.plotting import figure, show
from bokeh.palettes import Category20
from bokeh.io import output_notebook
import random

output_notebook()


copy_nfile=concrete_pluscat.copy()
type_group=copy_nfile['type_mix'].unique()

if len(type_group)<=2:
  col=[Category20[5][0],Category20[5][3]]
else:
  col=Category20[len(type_group)]


type_mix=sorted(copy_nfile['type_mix'].unique())
gf1=copy_nfile[copy_nfile['type_mix']==type_mix[0]]
gf2=copy_nfile[copy_nfile['type_mix']==type_mix[1]]
gf3=copy_nfile[copy_nfile['type_mix']==type_mix[2]]
gf4=copy_nfile[copy_nfile['type_mix']==type_mix[3]]
gf5=copy_nfile[copy_nfile['type_mix']==type_mix[4]]
gf6=copy_nfile[copy_nfile['type_mix']==type_mix[5]]

print(type_mix)
z=figure(title='strength vs age by mix_type',height=350)
z.scatter(x=sorted(gf1['age'].unique()),y=gf1.groupby('age')['strength'].mean(),color=col[0],legend_label=type_mix[0])
z.scatter(x=sorted(gf2['age'].unique()),y=gf2.groupby('age')['strength'].mean(),color=col[1],legend_label=type_mix[1])
z.scatter(x=sorted(gf3['age'].unique()),y=gf3.groupby('age')['strength'].mean(),color=col[2],legend_label=type_mix[2])
z.scatter(x=sorted(gf4['age'].unique()),y=gf4.groupby('age')['strength'].mean(),color=col[3],legend_label=type_mix[3])
z.scatter(x=sorted(gf5['age'].unique()),y=gf5.groupby('age')['strength'].mean(),color=col[4],legend_label=type_mix[4])
z.scatter(x=sorted(gf6['age'].unique()),y=gf6.groupby('age')['strength'].mean(),color=col[5],legend_label=type_mix[5])
z.xaxis.axis_label='age (days)'
z.yaxis.axis_label='strength (Mpa)'
z.add_layout(z.legend[0], 'right')
z.background_fill_color = "#eff0e7"
show(z)

['all mix', 'ash', 'ash-superplastic', 'just cement', 'slag', 'slag-superplastic', 'superplastic']


The dummies for the categorical values were created so we can added to the model later:

In [37]:
# Create the dataframe with all the numerical and categorical values
concrete_data=pd.get_dummies(data=concrete_pluscat,columns=['type_mix'],drop_first=True).astype(int)
concrete_data.head()


cement  slag  ash  water  superplastic  coarseagg  fineagg  age  strength  \
0     141   212    0    203             0        971      748   28        29   
1     168    42  124    158            10       1080      796   14        23   
2     250     0   95    187             5        956      861   28        29   
3     266   114    0    228             0        932      670   28        45   
4     154   183    0    193             9       1047      696   28        18   

   type_mix_ash  type_mix_ash-superplastic  type_mix_just cement  \
0             0                          0                     0   
1             0                          0                     0   
2             0                          1                     0   
3             0                          0                     0   
4             0                          0                     0   

   type_mix_slag  type_mix_slag-superplastic  type_mix_superplastic  
0              1                           0                      0  
1              0                           0                      0  
2              0                           0                      0  
3              1                           0                      0  
4              0                           1                      0

In [38]:
RED = "\033[31m"
RESET = "\033[0m"

ptg_data28=(len(concrete[concrete['age']<=100])/len(concrete))*100


print(f'The percentage of data under 100 days over the total is {RED}{round(ptg_data28,3)} %{RESET}')

print(f"{RED}{len(concrete[concrete['age']>100])}{RESET} samples are over 100 days")
print(f"{RED}{len(concrete[concrete['age']<=100])}{RESET} samples are under 100 days")

The percentage of data under 100 days over the total is 93.981 %
62 samples are over 100 days
968 samples are under 100 days


in the first capstone project, it is possible to identify that big amount of the data is below 100 days, using mathematical operations we can observe that ***93,981 %*** of the samples are obtain before the 100 days, so this data is the most representative of the whole dataset, being the samples above that age potential noise in the model as they also represent values of just two particular mixes (slag and all mix). Cause of this, the samples above 100 days will be removed of the model an this will predict compressive strength between 0 and 100 days.

### Creation of the taining and testing samples
 The samples are split in testing and training data for dependent and independent variables using 'train_test_split' method. so we can use this data for further modeling, modeling over two different set of data:

    * Full data
    * Data filtered by age (age under 100 days)


In [39]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

X=concrete_data.drop(columns=['strength'])
y=concrete_data['strength']

X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25)


In [40]:
# concrete_model_data=concrete_data[concrete_data['age']<=100]

concrete_model_data=concrete_data[concrete_data['age']<=100]
X_100=concrete_model_data.drop(columns=['strength'])
y_100=concrete_model_data['strength']
X_train_100,X_test_100,y_train_100,y_test_100=train_test_split(X_100,y_100,test_size=0.25)
print(y_train_100.shape,X_train_100.shape,y_test_100.shape)

(726,) (726, 14) (242,)


In [41]:
# Seprating my numeric variables and nominal/categorical values
num_col=[a for a in concrete.columns]
num_col.remove('strength')
#numerical values
X_train_num=X_train[num_col]
X_test_num=X_test[num_col]
print(X_train_num.shape)
#numerical values dataset 2 (100 days)
X_train_num_100=X_train_100[num_col]
X_test_num_100=X_test_100[num_col]
print(X_train_num_100.shape)

#nominal values
X_train_nom=X_train.drop(columns=num_col)
X_test_nom=X_test.drop(columns=num_col)
#nominal values dataset 2 (100 days)
X_train_nom_100=X_train_100.drop(columns=num_col)
X_test_nom_100=X_test_100.drop(columns=num_col)

print(X_train.shape)

(772, 8)
(726, 8)
(772, 14)


##Scalling the data

In [42]:
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler, MinMaxScaler

scale_data=StandardScaler()
print(pd.DataFrame(scale_data.fit_transform(X_train_num)))

X_train_num_scale=scale_data.fit_transform(X_train_num)
X_test_num_scale=scale_data.fit_transform(X_test_num)

#dataset 2 (100 days)
X_train_num_scale_100=scale_data.fit_transform(X_train_num_100)
X_test_num_scale_100=scale_data.fit_transform(X_test_num_100)


            0         1         2         3         4         5         6  \
0   -1.277555  1.188987 -0.835732 -0.454053 -0.674913  0.324827  0.663214   
1   -0.743400  2.712976 -0.835732  1.029863 -1.020540 -0.152430 -1.770381   
2    1.412640  0.380101 -0.835732 -1.381500  2.090103 -0.500699  0.355942   
3    2.383830 -0.862536 -0.835732  0.380650 -1.020540  1.937183 -1.979326   
4   -0.044144  1.247602  1.038489 -0.871404  0.707595 -1.867977 -0.111112   
..        ...       ...       ...       ...       ...       ...       ...   
767 -1.093029  0.649730  1.163437 -0.268563  0.189155  0.402220 -0.344638   
768 -1.258131 -0.862536  2.053692 -0.685915  1.571662  1.279342 -1.180419   
769 -1.423233  1.482062 -0.835732  0.519767 -1.020540 -0.616789  0.798414   
770  0.674536 -0.862536 -0.835732  0.519767 -1.020540  1.047162  0.429687   
771 -1.277555 -0.862536  2.006836  0.009671  1.571662 -1.751887  1.351504   

            7  
0   -0.278424  
1   -0.655385  
2   -0.278424  
3    3.3705

###Lineal model

In [43]:
from sklearn import linear_model
#using all data
lineal=linear_model.LinearRegression()
lineal.fit(X_train_num_scale,y_train)
l0=round(lineal.score(X_train_num_scale,y_train),3)
l1=round(lineal.score(X_test_num_scale,y_test),3)
score_lineal=[l0,l1]
#dataset 2 (100 days)

lineal_100=linear_model.LinearRegression()
lineal_100.fit(X_train_num_scale_100,y_train_100)
la=round(lineal_100.score(X_train_num_scale_100,y_train_100),3)
lb=round(lineal_100.score(X_test_num_scale_100,y_test_100),3)
score_lineal_100=[la,lb]
df_lineal=pd.DataFrame({'type_score':['Training','Testing'],'all data R^2':score_lineal,'filtered data R^2':score_lineal_100}).set_index('type_score')
print('values of scores for lineal model')
print(df_lineal)

values of scores for lineal model
            all data R^2  filtered data R^2
type_score                                 
Training           0.606              0.771
Testing            0.626              0.724


###Reciprocal logaritmic


In [44]:
import numpy as np

def addReciprocalLogFeatures(numeric):
    log_feats = numeric.copy()
    valid = (log_feats != 1) & (log_feats > 0)
    log_feats[valid] = np.log(log_feats[valid]) / np.log(10)
    log_feats[log_feats <= 0] = 1e-10
    rec_log_feats = log_feats
    return np.hstack([numeric, rec_log_feats, numeric * rec_log_feats])

X_train_reclog = addReciprocalLogFeatures(X_train_num_scale)
X_test_reclog = addReciprocalLogFeatures(X_test_num_scale)

#dataset 2 (100 days)

X_train_reclog_100 = addReciprocalLogFeatures(X_train_num_scale_100)
X_test_reclog_100 = addReciprocalLogFeatures(X_test_num_scale_100)


In [45]:
#log regression
lineal_rl=linear_model.LinearRegression()
lineal_rl.fit(X_train_reclog,y_train)
lr0=lineal_rl.score(X_train_reclog,y_train)
lr1=lineal_rl.score(X_test_reclog,y_test)
score_lr=[lr0,lr1]
#dataset 2 (100 days)
lineal_rl_100=linear_model.LinearRegression()
lineal_rl_100.fit(X_train_reclog_100,y_train_100)
lr0_100=lineal_rl_100.score(X_train_reclog_100,y_train_100)
lr1_100=lineal_rl_100.score(X_test_reclog_100,y_test_100)
score_lr_100=[lr0_100,lr1_100]
predicted_reclog_100=lineal_rl_100.predict(X_test_reclog_100)

df_lr=pd.DataFrame({'type_score':['Training','Testing'],'all data R^2':score_lr,'filtered data R^2':score_lr_100}).set_index('type_score')
print('values of scores for lineal model')
print(df_lr)

values of scores for lineal model
            all data R^2  filtered data R^2
type_score                                 
Training        0.785392           0.854532
Testing         0.770762           0.796344


In [46]:
BLUE = "\033[1;34m"
RESET = "\033[0m"

df1 = pd.DataFrame({"Predicted_100": predicted_reclog_100, "Actual_100": y_test_100})
df1['% Difference(100 days)'] = (abs(df1['Predicted_100']-df1['Actual_100'])/df1['Actual_100'])*100

print("Percentage Difference between Predicted and Actual Values (rec.log model with age bellow 100 days)")
print(df1.head())
print(f"\nMean % Difference between Predicted and Actual Values of data bellow 100 days:{BLUE}{str(round(df1['% Difference(100 days)'].mean(),3))}%{RESET}")

Percentage Difference between Predicted and Actual Values (rec.log model with age bellow 100 days)
     Predicted_100  Actual_100  % Difference(100 days)
514      50.819240          51                0.354431
811      44.987772          59               23.749539
730      44.313873          50               11.372254
413       6.128861          10               38.711387
49       40.048147          33               21.358022

Mean % Difference between Predicted and Actual Values of data bellow 100 days:20.417%


###cubic model

In [47]:
#lineal regession gives a r2 too low and the function seems to fit to a log function
from sklearn.preprocessing import PolynomialFeatures,FunctionTransformer

from sklearn.pipeline import make_pipeline

square = PolynomialFeatures(degree=3, include_bias=False)
X_train_scale_cubic = square.fit_transform(X_train_num_scale)
X_test_scale_cubic = square.fit_transform(X_test_num_scale)

reg=linear_model.LinearRegression()
reg.fit(X_train_scale_cubic,y_train)
prediction_train=reg.predict(X_train_scale_cubic)
prediction_test=reg.predict(X_test_scale_cubic)

C0=reg.score(X_train_scale_cubic,y_train)
C1=reg.score(X_test_scale_cubic,y_test)
cubic_score=[C0,C1]

#filtered data
cubic_f=PolynomialFeatures(degree=3, include_bias=False)
X_train_scale_cubic_100 = cubic_f.fit_transform(X_train_num_scale_100)
X_test_scale_cubic_100 = cubic_f.fit_transform(X_test_num_scale_100)

C0_100=reg.score(X_train_scale_cubic_100,y_train_100)
C1_100=reg.score(X_test_scale_cubic_100,y_test_100)
cubic_score_100=[C0_100,C1_100]
Predict_cubic=reg.predict(X_test_scale_cubic)

df_cubic=pd.DataFrame({'type_score':['Training','Testing'],'All data R^2':cubic_score,'Filtered data R^2':cubic_score_100}).set_index('type_score')
print('values of scores for cubic model')
print(df_cubic)

values of scores for cubic model
            All data R^2  Filtered data R^2
type_score                                 
Training        0.930068          -1.029224
Testing         0.655544          -0.783196


In [48]:
df_cubic = pd.DataFrame({"Predicted": Predict_cubic, "Actual": y_test})
df_cubic['% Difference(all data)'] = (abs(df_cubic['Predicted']-df_cubic['Actual'])/df_cubic['Actual'])*100

print("Percentage Difference between Predicted and Actual Values (cubic model)")
print(df_cubic.head())
print(f"\nMean % Difference between Predicted and Actual Values with all data:{BLUE}{str(round(df_cubic['% Difference(all data)'].mean(),3))} %{RESET}")


Percentage Difference between Predicted and Actual Values (cubic model)
     Predicted  Actual  % Difference(all data)
711  40.333399      43                6.201397
991  67.217347      75               10.376870
714  31.355236      34                7.778718
194   8.766512      23               61.884729
684  17.921494      29               38.201746

Mean % Difference between Predicted and Actual Values with all data:20.3 %


**Removing the data above 100 days improve the model consideraby  the scores in all 3 different models but the cubic polynomial one, which shows very negative values.**

## ***Adding nominal values***
  The two best models were selected and added the nominal values, so in this way will be possible to asses their performance.
  
  * Cubic model with all data
  * rec.log model with filtered data

In [49]:

X_train_ln_full = np.hstack([X_train_scale_cubic, X_train_nom])
X_test_ln_full = np.hstack([X_test_scale_cubic, X_test_nom])
#filtered data
X_train_ln_full_100 = np.hstack([X_train_reclog_100, X_train_nom_100])
X_test_ln_full_100 = np.hstack([X_test_reclog_100, X_test_nom_100])
X_train_ln_full.shape

(772, 170)

In [50]:
RED = "\033[1;31m"
GREEN= "\033[1;32m"
RESET = "\033[0m"

ln_model= linear_model.RidgeCV()
ln_model.fit(X_train_ln_full,y_train)
train_predict_ln=ln_model.predict(X_train_ln_full)
test_predict_ln=ln_model.predict(X_test_ln_full)
print('Considering the nominal values for the two best models in their respective scenarios')
print('Cubic model with all the data')
print(f'For the train data,the R^2 for the {RED}cubic model{RESET} is : {round(ln_model.score(X_train_ln_full,y_train),3)}')
print(f'For the test data,the R^2 for the {RED}cubic model{RESET}is : {round(ln_model.score(X_test_ln_full,y_test),3)}')

#filtered model

ln_model_100= linear_model.RidgeCV()
ln_model_100.fit(X_train_ln_full_100,y_train_100)
train_predict_ln_100=ln_model_100.predict(X_train_ln_full_100)
test_predict_ln_100=ln_model_100.predict(X_test_ln_full_100)

print('logaritmic model filtered (bellow 11 days)')
print(f'For the train data,the R^2 for the {GREEN}reciprocal log model{RESET} is : {round(ln_model_100.score(X_train_ln_full_100,y_train_100),3)}')
print(f'For the test data,the R^2 for the {GREEN}reciprocal log model{RESET} is : {round(ln_model_100.score(X_test_ln_full_100,y_test_100),3)}')

Considering the nominal values for the two best models in their respective scenarios
Cubic model with all the data
For the train data,the R^2 for the cubic model is : 0.925
For the test data,the R^2 for the cubic modelis : 0.736
logaritmic model filtered (bellow 11 days)
For the train data,the R^2 for the reciprocal log model is : 0.855
For the test data,the R^2 for the reciprocal log model is : 0.801


In [51]:
df0 = pd.DataFrame({"Predicted": test_predict_ln, "Actual": y_test})
df1 = pd.DataFrame({"Predicted_100": test_predict_ln_100, "Actual_100": y_test_100})
df0['% Difference(all data)'] = (abs(df0['Predicted']-df0['Actual'])/df0['Actual'])*100
df1['% Difference(100 days)'] = (abs(df1['Predicted_100']-df1['Actual_100'])/df1['Actual_100'])*100

print("Percentage Difference between Predicted and Actual Values (cubic model)")
print(df0.head())
print('----------------------------------------------')
print("Percentage Difference between Predicted and Actual Values (rec.log model with age bellow 100 days)")
print(df1.head())
print(f"\nMean % Difference between Predicted and Actual Values with all data:{RED}{str(round(df0['% Difference(all data)'].mean(),3))} %{RESET}")
print(f"\nMean % Difference between Predicted and Actual Values of data bellow 100 days:{GREEN}{str(round(df1['% Difference(100 days)'].mean(),3))}%{RESET}")

Percentage Difference between Predicted and Actual Values (cubic model)
     Predicted  Actual  % Difference(all data)
711  41.661330      43                3.113186
991  68.772398      75                8.303469
714  30.996388      34                8.834153
194  12.824470      23               44.241433
684  12.630548      29               56.446386
----------------------------------------------
Percentage Difference between Predicted and Actual Values (rec.log model with age bellow 100 days)
     Predicted_100  Actual_100  % Difference(100 days)
514      49.809714          51                2.333895
811      45.704981          59               22.533930
730      44.125687          50               11.748626
413       7.980014          10               20.199858
49       40.394644          33               22.408013

Mean % Difference between Predicted and Actual Values with all data:18.075 %

Mean % Difference between Predicted and Actual Values of data bellow 100 days:20.462%


In [52]:
results=[]
results2=[]
n_runs=10
for i in range(n_runs):
  #Split all data set

  X=concrete_data.drop(columns=['strength'])
  y=concrete_data['strength']
  X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.25)

  # split the filtered data
  concrete_model_data=concrete_data[concrete_data['age']<=100]
  X_100=concrete_model_data.drop(columns=['strength'])
  y_100=concrete_model_data['strength']
  X_train_100,X_test_100,y_train_100,y_test_100=train_test_split(X_100,y_100,test_size=0.25)


  # Seprating my numeric variables and nominal/categorical values
  num_col=[a for a in concrete.columns]
  num_col.remove('strength')
  #numerical values
  X_train_num=X_train[num_col]
  X_test_num=X_test[num_col]
  #nominal values
  X_train_nom=X_train.drop(columns=num_col)
  X_test_nom=X_test.drop(columns=num_col)

  #numerical values dataset 2 (100 days)
  X_train_num_100=X_train_100[num_col]
  X_test_num_100=X_test_100[num_col]
  #nominal values dataset 2 (100 days)
  X_train_nom_100=X_train_100.drop(columns=num_col)
  X_test_nom_100=X_test_100.drop(columns=num_col)
  #------------------------------------------------
  #scalling data
  scale_data=StandardScaler()
  X_train_num_scale=scale_data.fit_transform(X_train_num)
  X_test_num_scale=scale_data.fit_transform(X_test_num)
  #dataset 2 (100 days)
  X_train_num_scale_100=scale_data.fit_transform(X_train_num_100)
  X_test_num_scale_100=scale_data.fit_transform(X_test_num_100)
#Modeling the datasets
#Cubic model all data
  square = PolynomialFeatures(degree=3, include_bias=False)
  X_train_scale_cubic = square.fit_transform(X_train_num_scale)
  X_test_scale_cubic = square.fit_transform(X_test_num_scale)
  X_train_ln_full = np.hstack([X_train_scale_cubic, X_train_nom])
  X_test_ln_full = np.hstack([X_test_scale_cubic, X_test_nom])

  ln_model= linear_model.RidgeCV()
  ln_model.fit(X_train_ln_full,y_train)
  #calculating R^2
  train_score_ln=round(ln_model.score(X_train_ln_full,y_train),3)
  test_score_ln=round(ln_model.score(X_test_ln_full,y_test),3)
  #calculating variance in prediction and real values mean
  train_predict_ln=ln_model.predict(X_train_ln_full)
  test_predict_ln=ln_model.predict(X_test_ln_full)

  df0 = pd.DataFrame({"Predicted": test_predict_ln, "Actual": y_test})
  df0['% Difference(all data)'] = (abs(df0['Predicted']-df0['Actual'])/df0['Actual'])*100
  mean_valpred_cubic=str(round(df0['% Difference(all data)'].mean(),3))

  results.append({'run':i+1,
                  'R^2 train':train_score_ln,
                  'R^2 test': test_score_ln,
                  'Mean Variance(real/prediction)':mean_valpred_cubic,
                  'intercept':ln_model.intercept_,
                  'Coefficients':ln_model.coef_})
  #Reciprocal logaritmic model

  X_train_reclog_100 = addReciprocalLogFeatures(X_train_num_scale_100)
  X_test_reclog_100 = addReciprocalLogFeatures(X_test_num_scale_100)
  X_train_ln_full_100 = np.hstack([X_train_reclog_100, X_train_nom_100])
  X_test_ln_full_100 = np.hstack([X_test_reclog_100, X_test_nom_100])
  ln_model_100= linear_model.RidgeCV()
  ln_model_100.fit(X_train_ln_full_100,y_train_100)
  #results values of filtered data
  test_predict_ln_100=ln_model_100.predict(X_test_ln_full_100)
  df1 = pd.DataFrame({"Predicted_100": test_predict_ln_100, "Actual_100": y_test_100})
  df1['% Difference(100 days)'] = (abs(df1['Predicted_100']-df1['Actual_100'])/df1['Actual_100'])*100
  mean_valpred_ln=str(round(df1['% Difference(100 days)'].mean(),3))

  train_score_ln_100=round(ln_model_100.score(X_train_ln_full_100,y_train_100),3)
  test_score_ln_100=round(ln_model_100.score(X_test_ln_full_100,y_test_100),3)

  results2.append({'run':i+1,
                  'R^2 train':train_score_ln_100,
                  'R^2 test': test_score_ln_100,
                  'Mean Variance(real/prediction)':mean_valpred_ln,
                  'intercept':ln_model_100.intercept_,
                  'Coefficients':ln_model_100.coef_})
# Convert results to a DataFrame
results_model = pd.DataFrame(results)
results_model_100 = pd.DataFrame(results2)

# Split coefficients into separate columns for better readability
# coefficients_df = pd.DataFrame(results_model['Coefficients'].tolist(), columns=[f'Coefficient_{i+1}' for i in range(X.shape[1])])
# final_model_results = pd.concat([results_model.drop(columns='Coefficients'), coefficients_df], axis=1)

# Display results
print(results_model)
print(results_model_100)


   run  R^2 train  R^2 test Mean Variance(real/prediction)  intercept  \
0    1      0.930     0.828                         23.518  45.151484   
1    2      0.913     0.798                         22.509  43.440343   
2    3      0.931     0.777                         17.048  44.518777   
3    4      0.927     0.849                         17.881  44.394705   
4    5      0.925     0.858                          17.15  43.584889   
5    6      0.925     0.793                         19.947  43.165292   
6    7      0.922     0.887                         16.789  44.507179   
7    8      0.926     0.814                         18.196  44.020253   
8    9      0.927     0.843                         22.146  44.215075   
9   10      0.927     0.843                         16.528  45.280124   

                                        Coefficients  
0  [6.628872170987295, 4.614525434789613, -1.3073...  
1  [4.448791781073643, 3.3598262603871523, -0.198...  
2  [6.631874015928694, 2.996448

It is possible to observe that 2 of the 3 models improved narrowing the data to just 968 samples , being the cubic model negativly affected by this reduction on the data but clearly showing signs of overfitting in the predictive model as the difference between the R^2 for train and test data is above 5% difference in most of the cases, reaching in some of them 10%, however with the reciprocal logaritmic model, the reduction in the data gives us  slightly lower R^2 for the test data but a difference between the coeficient of determination in between 1 and 2%.


## applying PCA

In [53]:
X_train_WS=X_train[['water','superplastic']]
X_train_WS_scaled= scale_data.fit_transform(X_train_WS)
X_test_WS=X_test[['water','superplastic']]

X_test_WS_scaled= scale_data.fit_transform(X_test_WS)
print(X_train_WS_scaled.shape,X_test_WS_scaled.shape)

pca=PCA(n_components=1)
X_train_pca=pca.fit_transform(X_train_WS_scaled)
X_test_pca=pca.fit_transform(X_test_WS_scaled)

# # train and test data without water and superplastic

col=[a for a in concrete.columns]
items_to_remove = ['strength','water','superplastic']
col = [c for c in col if c not in items_to_remove]

X_train_non_PCA=X_train[col]
X_test_non_PCA=X_test[col]
#scaled data
X_train_non_PCA_scale=scale_data.fit_transform(X_train_non_PCA)
X_test_non_PCA_scale=scale_data.fit_transform(X_test_non_PCA)

X_train_PCA_scaled=np.hstack([X_train_pca,X_train_non_PCA_scale])
X_test_PCA_scaled=np.hstack([X_test_pca,X_test_non_PCA_scale])


(772, 2) (258, 2)


In [54]:
X_train_reclog_PCA = addReciprocalLogFeatures(X_train_PCA_scaled)
X_test_reclog_PCA = addReciprocalLogFeatures(X_test_PCA_scaled)

reg=linear_model.LinearRegression()
reg.fit(X_train_reclog_PCA,y_train)
prediction_train=reg.predict(X_train_reclog_PCA)
prediction_test=reg.predict(X_test_reclog_PCA)

C0=reg.score(X_train_reclog_PCA,y_train)
C1=reg.score(X_test_reclog_PCA,y_test)
PCA_score=[C0,C1]
df_PCA=pd.DataFrame({'type_score':['Training','Testing'],'All data R^2':PCA_score}).set_index('type_score')
print('values of scores for PCA model')
print(df_PCA)


values of scores for PCA model
            All data R^2
type_score              
Training        0.764554
Testing         0.786908


In this case the R^2 for the train data was reduced but the R^2 for the test data increased in relation with the previous model.
 So, now left to add the nominal variabes to verify the score with this model.

In [55]:

X_train_ln_full = np.hstack([X_train_reclog_PCA, X_train_nom])
X_test_ln_full = np.hstack([X_test_reclog_PCA, X_test_nom])


In [56]:
ln_model= linear_model.RidgeCV()
ln_model.fit(X_train_ln_full,y_train)
train_predict_pca=ln_model.predict(X_train_ln_full)
test_predict_pca=ln_model.predict(X_test_ln_full)
print(ln_model.score(X_train_ln_full,y_train))
print(ln_model.score(X_test_ln_full,y_test))

0.7747694065466675
0.7949167681083145


In [57]:
df = pd.DataFrame({"Predicted": test_predict_pca, "Actual": y_test})
df['% Difference'] = (abs(df['Predicted']-df['Actual'])/df['Actual'])*100

print("Percentage Difference between Predicted and Actual Values (cubic model)")
print(df.head())
print("\nMean % Difference between Predicted and Actual Values: " + str(round(df['% Difference'].mean(),3)) +"%")

Percentage Difference between Predicted and Actual Values (cubic model)
     Predicted  Actual  % Difference
445  52.123307      40     30.308268
500  52.090824      49      6.307803
546  30.905906      37     16.470523
226  40.352421      51     20.877607
570  38.616016      50     22.767968

Mean % Difference between Predicted and Actual Values: 20.932%


# Conclusion

  The best performance was obtain with the filtered data using the reciprocal log model, without any benefit reducing the dimension of the data using PCA for water and superplastic. These values vary each time we ran the models, but the filtered data consistently performed better than the full data set (except the cubic model). By the other hand, adding nominal data slightly improve performance.
  Even tough the cubic model ocasionally shows a slightly better perfomrance than the recirpocal logarithmic model, it shows a clear overfitting, which makes the logarithmic model more reliable.